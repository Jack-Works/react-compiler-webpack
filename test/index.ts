import chai from 'chai';
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';

import { defineReactCompilerLoaderOption } from '../dist';

import getWebpackCompiler from './utils/get-webpack-compiler';
import getRspackCompiler from './utils/get-rspack-compiler';

import compile from './utils/compile';
import getModuleSource from './utils/get-module-source';

chai.should();
chai.use(jestSnapshotPlugin());

const defaultOption = defineReactCompilerLoaderOption({
  sources() { return true; }
});

([
  ['react-compiler-webpack (webpack)', getWebpackCompiler],
  ['react-compiler-webpack (rspack)', getRspackCompiler]
] as const).forEach(([name, getCompiler]) => {
  describe(name, () => {
    it('defineForgettiLoaderOptions', () => {
      const opt = {};
      const definedOpt = defineReactCompilerLoaderOption({});

      definedOpt.should.eql(opt);
    });

    it('should work', async () => {
      const [compiler, fs] = getCompiler('./simple.jsx', defaultOption);
      const stats = await compile(compiler);

      getModuleSource('./simple.jsx', stats, fs)?.should.toMatchSnapshot('simple.jsx');
    });

    it('should work with tsx', async () => {
      const [compiler, fs] = getCompiler('./simple.tsx', defaultOption);
      const stats = await compile(compiler);

      getModuleSource('./simple.tsx', stats, fs)?.should.toMatchSnapshot('simple.tsx');
    });

    it('should optimize complex component', async () => {
      const [compiler, fs] = getCompiler('./complex.tsx', defaultOption);
      const stats = await compile(compiler);

      getModuleSource('./complex.tsx', stats, fs)?.should.toMatchSnapshot('complex.tsx');
    });
  });
});
