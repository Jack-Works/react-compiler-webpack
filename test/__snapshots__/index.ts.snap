// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-compiler-webpack (webpack) should optimize complex component 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	var __webpack_modules__ = ({

/***/ \\"./complex.tsx\\":
/*!*********************!*\\\\
  !*** ./complex.tsx ***!
  \\\\*********************/
/***/ (() => {

throw new Error(\\"Module build failed (from ../../dist/react-compiler-loader.js):\\\\nReactCompilerError: InvalidConfig: Expected a filename but found none.. When the 'sources' config options is specified, the React compiler will only compile files with a name\\\\n    at compileProgram (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/babel-plugin-react-compiler@0.0.0-experimental-c23de8d-20240515/node_modules/babel-plugin-react-compiler/dist/index.js:166970:21)\\\\n    at PluginPass.call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/babel-plugin-react-compiler@0.0.0-experimental-c23de8d-20240515/node_modules/babel-plugin-react-compiler/dist/index.js:167511:9)\\\\n    at call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/visitors.ts:293:14)\\\\n    at NodePath._call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:35:20)\\\\n    at NodePath.call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:20:17)\\\\n    at NodePath.visit (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:94:31)\\\\n    at TraversalContext.visitQueue (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:147:16)\\\\n    at TraversalContext.visitSingle (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:108:19)\\\\n    at TraversalContext.visit (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:179:19)\\\\n    at traverseNode (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/traverse-node.ts:40:17)\\\\n    at traverse (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/index.ts:82:15)\\\\n    at transformFile (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transformation/index.ts:124:15)\\\\n    at transformFile.next (<anonymous>)\\\\n    at run (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transformation/index.ts:48:12)\\\\n    at run.next (<anonymous>)\\\\n    at transform (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transform.ts:29:20)\\\\n    at transform.next (<anonymous>)\\\\n    at step (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:261:32)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:273:13\\\\n    at async.call.result.err.err (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:223:11)\\\\n    at cb (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:189:28)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/gensync-utils/async.ts:90:7\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:113:33\\\\n    at step (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:287:14)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:273:13\\\\n    at async.call.result.err.err (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:223:11)\\");

/***/ })

/******/ 	});
/************************************************************************/
/******/ 	
/******/ 	// startup
/******/ 	// Load entry module and return exports
/******/ 	// This entry module doesn't tell about it's top-level declarations so it can't be inlined
/******/ 	var __webpack_exports__ = {};
/******/ 	__webpack_modules__[\\"./complex.tsx\\"]();
/******/ 	
/******/ })()
;"
`;

exports[`react-compiler-webpack (webpack) should work 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	var __webpack_modules__ = ({

/***/ \\"./simple.jsx\\":
/*!********************!*\\\\
  !*** ./simple.jsx ***!
  \\\\********************/
/***/ (() => {

throw new Error(\\"Module build failed (from ../../dist/react-compiler-loader.js):\\\\nReactCompilerError: InvalidConfig: Expected a filename but found none.. When the 'sources' config options is specified, the React compiler will only compile files with a name\\\\n    at compileProgram (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/babel-plugin-react-compiler@0.0.0-experimental-c23de8d-20240515/node_modules/babel-plugin-react-compiler/dist/index.js:166970:21)\\\\n    at PluginPass.call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/babel-plugin-react-compiler@0.0.0-experimental-c23de8d-20240515/node_modules/babel-plugin-react-compiler/dist/index.js:167511:9)\\\\n    at call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/visitors.ts:293:14)\\\\n    at NodePath._call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:35:20)\\\\n    at NodePath.call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:20:17)\\\\n    at NodePath.visit (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:94:31)\\\\n    at TraversalContext.visitQueue (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:147:16)\\\\n    at TraversalContext.visitSingle (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:108:19)\\\\n    at TraversalContext.visit (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:179:19)\\\\n    at traverseNode (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/traverse-node.ts:40:17)\\\\n    at traverse (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/index.ts:82:15)\\\\n    at transformFile (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transformation/index.ts:124:15)\\\\n    at transformFile.next (<anonymous>)\\\\n    at run (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transformation/index.ts:48:12)\\\\n    at run.next (<anonymous>)\\\\n    at transform (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transform.ts:29:20)\\\\n    at transform.next (<anonymous>)\\\\n    at step (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:261:32)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:273:13\\\\n    at async.call.result.err.err (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:223:11)\\\\n    at cb (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:189:28)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/gensync-utils/async.ts:90:7\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:113:33\\\\n    at step (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:287:14)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:273:13\\\\n    at async.call.result.err.err (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:223:11)\\");

/***/ })

/******/ 	});
/************************************************************************/
/******/ 	
/******/ 	// startup
/******/ 	// Load entry module and return exports
/******/ 	// This entry module doesn't tell about it's top-level declarations so it can't be inlined
/******/ 	var __webpack_exports__ = {};
/******/ 	__webpack_modules__[\\"./simple.jsx\\"]();
/******/ 	
/******/ })()
;"
`;

exports[`react-compiler-webpack (webpack) should work with tsx 1`] = `
"/******/ (() => { // webpackBootstrap
/******/ 	var __webpack_modules__ = ({

/***/ \\"./simple.tsx\\":
/*!********************!*\\\\
  !*** ./simple.tsx ***!
  \\\\********************/
/***/ (() => {

throw new Error(\\"Module build failed (from ../../dist/react-compiler-loader.js):\\\\nReactCompilerError: InvalidConfig: Expected a filename but found none.. When the 'sources' config options is specified, the React compiler will only compile files with a name\\\\n    at compileProgram (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/babel-plugin-react-compiler@0.0.0-experimental-c23de8d-20240515/node_modules/babel-plugin-react-compiler/dist/index.js:166970:21)\\\\n    at PluginPass.call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/babel-plugin-react-compiler@0.0.0-experimental-c23de8d-20240515/node_modules/babel-plugin-react-compiler/dist/index.js:167511:9)\\\\n    at call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/visitors.ts:293:14)\\\\n    at NodePath._call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:35:20)\\\\n    at NodePath.call (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:20:17)\\\\n    at NodePath.visit (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/path/context.ts:94:31)\\\\n    at TraversalContext.visitQueue (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:147:16)\\\\n    at TraversalContext.visitSingle (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:108:19)\\\\n    at TraversalContext.visit (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/context.ts:179:19)\\\\n    at traverseNode (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/traverse-node.ts:40:17)\\\\n    at traverse (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+traverse@7.24.5/node_modules/@babel/traverse/src/index.ts:82:15)\\\\n    at transformFile (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transformation/index.ts:124:15)\\\\n    at transformFile.next (<anonymous>)\\\\n    at run (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transformation/index.ts:48:12)\\\\n    at run.next (<anonymous>)\\\\n    at transform (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/transform.ts:29:20)\\\\n    at transform.next (<anonymous>)\\\\n    at step (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:261:32)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:273:13\\\\n    at async.call.result.err.err (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:223:11)\\\\n    at cb (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:189:28)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/@babel+core@7.24.5/node_modules/@babel/core/src/gensync-utils/async.ts:90:7\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:113:33\\\\n    at step (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:287:14)\\\\n    at /Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:273:13\\\\n    at async.call.result.err.err (/Users/sukka/Project/react-compiler-webpack/node_modules/.pnpm/gensync@1.0.0-beta.2/node_modules/gensync/index.js:223:11)\\");

/***/ })

/******/ 	});
/************************************************************************/
/******/ 	
/******/ 	// startup
/******/ 	// Load entry module and return exports
/******/ 	// This entry module doesn't tell about it's top-level declarations so it can't be inlined
/******/ 	var __webpack_exports__ = {};
/******/ 	__webpack_modules__[\\"./simple.tsx\\"]();
/******/ 	
/******/ })()
;"
`;
